README.md
# react-images-viewer

[![Build Status](https://travis-ci.org/guonanci/react-images-viewer.svg?branch=master)](https://travis-ci.org/guonanci/react-images-viewer)

<!-- [![Coverage Status](https://coveralls.io/repos/github/guonanci/react-images-viewer/badge.svg?branch=master)](https://coveralls.io/github/guonanci/react-images-viewer?branch=master) -->

[![jest](https://jestjs.io/img/jest-badge.svg)](https://github.com/facebook/jest)

A react library that view photos list easily, and a simple, responsive viewer component for displaying an array of images.

[中文文档](./README_CN.md)

## Quick start

```bash
# recommended
yarn add react-images-viewer
```

or

```bash
npm install react-images-viewer --save
```

```jsx
import React from "react";
import ImgsViewer from "react-images-viewer";

export default class Demo extends React.Component {
  render() {
    return (
      <ImgsViewer
        imgs={[
          { src: "http://example.com/img1.jpg" },
          { src: "http://example.com/img2.png" },
        ]}
        currImg={this.state.currImg}
        isOpen={this.state.viewerIsOpen}
        onClickPrev={this.gotoPrevious}
        onClickNext={this.gotoNext}
        onClose={this.closeViewer}
      />
    );
  }
}
```

## Demo & Example

Live Demo: [guonanci.github.io/react-images-viewer](https://guonanci.github.io/react-images-viewer)

```bash
yarn install
yarn
yarn start
```

Then open [`localhost:8000`](http://localhost:8000) in a browser.

### Using srcSet

Example using srcSet:

```jsx
<ImgsViewer
  imgs={IMG_SET}
  ...
/>
```

```js
const IMG_SET = [
  {
    src: "http://example.com/img1.svg",
    caption: "A forest",
    // As an array
    srcSet: [
      "http://example.com/img1_1024.jpg 1024w",
      "http://example.com/img1_800.jpg 800w",
      "http://example.com/img1_500.jpg 500w",
      "http://example.com/img1_320.jpg 320w",
    ],
  },
  {
    src: "http://example.com/img2.svg",
    // As a string
    srcSet:
      "http://example.com/img2_1024.jpg 1024w, http://example.com/img2_800.jpg 800w, http://example.com/img2_500.jpg 500w, http://example.com/img2_320.jpg 320w",
  },
];
```

## Options

| Property            | Type       | Default          | Description                                                                                                                     |
| :------------------ | :--------- | :--------------- | :------------------------------------------------------------------------------------------------------------------------------ |
| backdropCloseable   | `boolean`  | `false`          | Allow users to exit the viewer by clicking the backdrop                                                                         |
| closeBtnTitle       | `str`      | `undefined`      | Customize close esc title                                                                                                       |
| enableKeyboardInput | `boolean`  | `true`           | Supports keyboard input - <code>space, esc</code>, <code> arrow left, arrow up</code>, and <code>arrow right, arrow down</code> |
| currImg             | `number`   | `0`              | Required if you want to navigate the imgsViewer, The index of the image to display initially                                    |
| customControls      | arr        | `undefined`      | An array of elements to display as custom controls on the top of viewer                                                         |
| imgs                | arr        | `undefined`      | Required. Array of image objects, See img opts table below                                                                      |
| imgCountSeparator   | `str`      | `' / '`          | Customize separator in the image count                                                                                          |
| isOpen              | `boolean`  | `false`          | Required if you want to navigate the imgsViewer, Whether or not the viewer is displayed                                         |
| leftArrowTitle      | `str`      | `undefined`      | Customize of left arrow title                                                                                                   |
| onClickPrev         | `function` | `undefined`      | Required if you want to navigate the imgsViewer, and fired on request of the previous image                                     |
| onClickNext         | `function` | `undefined`      | Required if you want to navigate the imgsViewer, and fired on request of the next image                                         |
| onClose             | `function` | `undefined`      | Required if you want to close the imgsViewer, and handle closing of the viewer                                                  |
| onClickImg          | `function` | `undefined`      | Handle click on current image                                                                                                   |
| onClickThumbnail    | `function` | `undefined`      | Handle click on thumbnail                                                                                                       |
| preloadNextImg      | `boolean`  | `true`           | Whether to preload the next available image                                                                                     |
| rightArrowTitle     | `str`      | `undefined`      | Customize right arrow title                                                                                                     |
| showCloseBtn        | `boolean`  | `true`           | Optionally display a close 'X' button in top right corner                                                                       |
| showImgCount        | `boolean`  | `true`           | Optionally display image index, e.g., "2 of 20"                                                                                 |
| width               | `number`   | `1024`           | Maximum width of the carousel; defaults to 1024px                                                                               |
| spinnerDisabled     | `boolean`  | `false`          | Disable Spinner                                                                                                                 |
| spinner             | `function` | `DefaultSpinner` | Spinner component class                                                                                                         |
| spinnerColor        | `str`      | `'#fff'`         | Color of spinner                                                                                                                |
| spinnerSize         | `number`   | `50`             | Size of spinner                                                                                                                 |
| preventAutoScroll   | `boolean`  | `true`           | Determines whether auto-scrolling is prevented                                                                                  |

## Images Object

| Property | Type      | Default     | Description |
| :------- | :-------- | :---------- | :---------- | -------- |
| src      | `str`     | `undefined` | Required    |
| srcSet   | `string[] | string`     | `undefined` | Optional |
| caption  | `str`     | `undefined` | Optional    |
| alt      | `str`     | `undefined` | Optional    |


.github/HISTORY.md
# React-Images-Viewer

### v1.6.8 / 2022-01-03

- Update: fix issues: #28, #38

### v1.3.0 / 2018-07-24

- Update: Make README.md and README_ZN.md desciption more specific
- Fix: correct `package.json - main` field value

---

### v1.0.5 / 2018-07-24

- Basic, thumbnailed, themed usage all work well
- Feature: you can use `up, right, bottom, left, space, esc` to navigate quickly
- Component API is very brief and precise


src/Viewer.js
import PropTypes from 'prop-types'
import React, { Component, Fragment } from 'react'
import { StyleSheet, css } from 'aphrodite'
import ScrollLock from 'react-scrolllock'

import defaultTheme from './theme'
import Arrow from './components/Arrow'
import Container from './components/Container'
import Footer from './components/Footer'
import Header from '/components/Header'
import PaginatedThumbnails from '/components/PaginatedThumbnails'
import Portal from '/components/Portal'
import DefaultSpinner from '/components/Spinner'

import { bindFunctions, canUseDom, deepMerge } from './utils/util'

function normalizeSourceSet(data) {
  const sourceSet = data.srcSet || data.srcset

  if (Array.isArray(sourceSet)) {
    return sourceSet.join()
  }

  return sourceSet
}

class Viewer extends Component {
  constructor (props) {
    super(props)

    this.theme = deepMerge(defaultTheme, this.theme)
    this.classes = StyleSheet.create(deepMerge(defaultStyles, this.theme))
    this.state = { imgLoaded: false }

    bindFunctions.call(this, [
      'gotoNext',
      'gotoPrev',
      'closeBackdrop',
      'handleKeyboardInput',
      'handleImgLoaded'
    ])
  }
  getChildContext () {
    return {
      theme: this.theme
    }
  }
  componentDidMount () {
    if (this.props.isOpen) {
      if (this.props.enableKeyboardInput) {
        window.addEventListener('keyboard', this.handleKeyboardInput)
      }
      if (typeof this.props.currImg === 'number') {
        this.preloadImg(this.props.currImg, this.handleImgLoaded)
      }
    }
  }
  UNSAFE_componentWillReceiveProps (nextProps) {
    if (!canUseDom) return

    // always to preload imgs with both directions
    // then when user changs direction, img also show quickly
    if (nextProps.preloadNextImg) {
      const nextIdx = nextProps.currImg + 1
      const prevIdx = nextProps.currImg - 1
      this.preloadImg(prevIdx)
      this.preloadImg(nextIdx)
    }
    // preload currImg
    if (this.props.currImg !== nextProps.currImg || !this.props.isOpen && nextProps.isOpen) {
      const img = this.preloadImg(nextProps.currImg, this.handleImgLoaded)
      this.setState({ imgLoaded: img.complete })
    }

    // add/remove event listeners
    if (!this.props.isOpen && nextProps.isOpen && nextProps.enableKeyboardInput) {
      window.addEventListener('keyboard', this.handleKeyboardInput)
    }
    if (!nextProps.isOpen && nextProps.enableKeyboardInput) {
      window.removeEventListener('keydown', this.handleKeyboardInput)
    }
  }
  UNSAFE_componentWillUnmount () {
    if (this.props.enableKeyboardInput) {
      window.removeEventListener('keydown', this.handleKeyboardInput)
    }
  }

  // ====================
  // Methods
  // ====================

  preloadImg (idx, onload) {
    const data = this.props.imgs[idx]

    if(!data) return

    const img = new Image()
    const sourceSet = normalizeSourceSet(data)

    // Todo: add error handling for missing imgs
    img.onerror = onload
    img.onload = onload
    img.src = data.src

    if (sourceSet) img.srcset = sourceSet

    return img

  }
  gotoNext (event) {
    const { currImg, imgs } = this.props
    const { imgLoaded } = this.state

    if (!imgLoaded || currImg === (imgs.length - 1)) return

    if (event) {
      event.preventDefault()
      event.stopPropagation()
    }

    this.props.onClickNext()
  }
  gotoPrev (event) {
    const { currImg } = this.props
    const { imgLoaded } = this.state

    if (!imgLoaded || currImg === 0) return

    if (event) {
      event.preventDefault()
      event.stopPropagation()
    }

    this.props.onClickPrev()
  }
  closeBackdrop (event) {
    if (event.target.id === 'viewerBackdrop' || event.target.tagName === 'FIGURE') {
      this.props.onClose()
    }
  }
  handleKeyboardInput (event) {
    const { keyCode } = event
    if (keyCode === 37 || keyCode === 33) { // left, up
      this.gotoPrev(event)
      return true
    } else if (keyCode === 39 || keyCode === 34) { // right, down
      this.gotoNext(event)
      return true
    } else if (keyCode === 27 || keyCode === 32) { // esc, space
      this.props.onClose()
      return true
    }
    return false
  }
  handleImgLoaded () {
    this.setState({ imgLoaded: true })
  }

  // ====================
  // Renderers
  // ====================

  renderArrowPrev () {
    if (this.props.currImg === 0) return null

    return (
      <Arrow
        direction="left"
        icon="arrowLeft"
        onclick={this.gotoPrev}
        title={this.props.leftArrowTitle}
        type="button"
      />
    )
  }
  renderArrowNext () {
    if (this.props.currImg === (this.props.imgs.length - 1)) return null

    return (
      <Arrow
        direction="right"
        icon="arrowRight"
        onClick={this.gotoNext}
        title={this.props.rightArrowTitle}
        type="button"
      />
    )
  }
  renderDialog () {
    const { backdropCloseable, isOpen, showThumbnails, width, } = this.props

    const { imgLoaded } = this.state

    if (!isOpen) return <span key="closed" />

    const offsetThumbnails = showThumbnails ? this.theme.thumnail.size + this.theme.container.gutter.vertical : 0

    return (
      <Container
        key="open"
        onClick={backdropCloseable && this.closeBackdrop}
        onTouchEnd={backdropCloseable && this.closeBackdrop}
      >
        <Fragment>
          <div className={css(this.classes.content)} style={{ marginBottom: offsetThumbnails, maxWidth: width }}>
            {imgLoaded && this.renderHeader()}
            {this.renderImgs()}
            {this.renderSpinner()}
            {imgLoaded && this.renderFooter()}
          </div>
          {imgLoaded && this.renderThumbnails()}
          {imgLoaded && this.renderArrowPrev()}
          {imgLoaded && this.renderArrowNext()}
          {this.props.preventScroll && <ScrollLock />}
        </Fragment>
      </Container>
    )
  }
  renderImgs () {
    const { currImg, imgs, onClickImg, showThumbnails } = this.props

    const { imgLoaded } = this.state

    if (!imgs || !imgs.length) return null

    const img = imgs[currImg]
    const sourceSet = normalizeSourceSet(img)
    const sizes = sourceSet ? '100vw' : null

    const thumbnailsSize = showThumbnails ? this.theme.thumnail.size : 0
    const heightOffset = `${this.theme.header.height + this.theme.footer.height + thumbnailsSize + (this.theme.container.gutter.vertical)}px`

    return (
      <figure className={css(this.classes.figure)}>
        {/* Re-implement when react warning 'unknown props'
        https://fb.me/react-unknown-props is resolved
        <Swipeable onSwipedLeft={this.gotonext} onSwipedRight={this.gotoPrev} /> */}
        <img
          className={css(this.classes.img, imgLoaded && this.classes.imgLoaded)}
          onClick={onClickImg}
          sizes={sizes}
          alt={img.alt}
          src={img.src}
          srcSet={sourceSet}
          style={{
            cursor: onClickImg ? 'pointer' : 'auto',
            maxHeight: `calc(100vh - ${heightOffset}`,
          }}
        />
      </figure>
    )
  }
  renderThumbnails () {
    const { imgs, currImg, onClickThumbnail, showThumbnails, thumbnailOffset } = this.props

    if (!showThumbnails) return null

    return (
      <PaginatedThumbnails
        currImg={currImg}
        imgs={imgs}
        offset={thumbnailOffset}
        onClickThumnail={onClickThumbnail}
      />
    )
  }
  renderHeader () {
    const { closeBtnTitle, customControls, onClose, showCloseBtn } = this.props

    return (
      <Header
        customControls={customControls}
        onClose={onClose}
        showCloseBtn={showCloseBtn}
        closeBtnTitle={closeBtnTitle}
      />
    )
  }
  renderFooter () {
    const { currImg, imgs, imgCountSeparator, showImgCount } = this.props

    if (!imgs || !imgs.length) return null

    return (
      <Footer
        caption={imgs[currImg].caption}
        countCurr={currImg + 1}
        countSeparator={imgCountSeparator}
        countTotal={imgs.length}
        showCount={showImgCount}
      />
    )
  }
  renderSpinner () {
    const { spinner, spinnerColor, spinnerSize } = this.props

    const { imgLoaded } = this.state
    const Spinner = spinner

    return (
      <div className={css(this.classes.spinner, !imgLoaded && this.classes.spinnerActive)}>
        <Spinner
          color={spinnerColor}
          size={spinnerSize}
        />
      </div>
    )
  }
  render () {
    return (
      <Portal>
        {this.renderDialog()}
      </Portal>
    )
  }
}

Viewer.propTypes = {
  backdropCloseable: PropTypes.bool,
  closeBtnTitle: PropTypes.string,
  currImg: PropTypes.number,
  customControls: PropTypes.arrayOf(PropTypes.node),
  enableKeyboardInput: PropTypes.bool,
  imgCountSeparator: PropTypes.string,
  imgs: PropTypes.arrayOf(
    PropTypes.shape({
      src: PropTypes.string.isRequired,
      srcSet: PropTypes.array,
      caption: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
      thumnail: PropTypes.string
    })
  ).isRequired,
  isOpen: PropTypes.bool,
  leftArrowTitle: PropTypes.string,
  onClickImg: PropTypes.func,
  onClickNext: PropTypes.func,
  onClickPrev: PropTypes.func,
  onClickThumbnail: PropTypes.func,
  onClose: PropTypes.func.isRequired,
  preloadNextImg: PropTypes.bool,
  preventScroll: PropTypes.bool,
  rightArrowTitle: PropTypes.string,
  showCloseBtn: PropTypes.bool,
  showImgCount: PropTypes.bool,
  showThumbnails: PropTypes.bool,
  spinner: PropTypes.func,
  spinnerColor: PropTypes.string,
  spinnerSize: PropTypes.number,
  theme: PropTypes.object,
  thumbnailOffset: PropTypes.number,
  width: PropTypes.number,
}
Viewer.defaultProps = {
  closeBtnTitle: '关闭（空格）',
  currImg: 0,
  enableKeyboardInput: true,
  imgCountSeparator: ' / ',
  leftArrowTitle: '上一张',
  onClickShowNextImg: true,
  preloadNextImg: true,
  preventScroll: true,
  rightArrowTitle: '下一张',
  showCloseBtn: true,
  showImgCount: true,
  spinner: DefaultSpinner,
  spinnerColor: '#fff',
  spinnerSize: 100,
  theme: {},
  thumbnailOffset: 2,
  width: 1024,
}
Viewer.childContextTypes = {
  theme: PropTypes.object.isRequired,
}

const defaultStyles = {
  content: {
    position: 'relative',
  },
  figure: {
    margin: 0, // remove browser default
  },
  img: {
    display: 'block', // removes browser default gutter
    height: 'auto',
    margin: '0 auto', // main center on very short screens or very narrow img
    maxWidth: '100%',

    // disable user select
    WebkitTouchCallout: 'none',
    userSelect: 'none',

    // opacity animation on image load
    opacity: 0,
    transition: 'opacity .3s',
  },
  imgLoaded: {
    opcaty: 1,
  },
  spinner: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',

    // opacity animation to make spinner appear with delay
    opacity: 0,
    transition: 'opacity .3s',
  },
  spinnerActive: {
    opacity: 1,
  }
}

export default Viewer
